src/corehost/cli/hostpolicy/breadcrumbs.cpp

```diff
@@ -8,54 +8,43 @@
 #include <trace.h>
 #include "breadcrumbs.h"
 
-breadcrumb_writer_t::breadcrumb_writer_t(bool enabled, const std::unordered_set<pal::string_t> &files)
-    : m_files(files)
-    , m_enabled(enabled)
-    , m_status(false)
+breadcrumb_writer_t::breadcrumb_writer_t(std::unordered_set<pal::string_t> &files)
 {
-    if (enabled && !pal::get_default_breadcrumb_store(&m_breadcrumb_store))
+    assert(m_files.empty());
+    m_files.swap(files);
+    assert(files.empty());
+    if (!pal::get_default_breadcrumb_store(&m_breadcrumb_store))
     {
         m_breadcrumb_store.clear();
     }
 }
 
-breadcrumb_writer_t::~breadcrumb_writer_t()
-{
-    if (m_enabled)
-    {
-        end_write();
-    }
-}
-
 // Begin breadcrumb writing: write synchronously or launch a
 // thread to write breadcrumbs.
-void breadcrumb_writer_t::begin_write()
+std::shared_ptr<breadcrumb_writer_t> breadcrumb_writer_t::begin_write(std::unordered_set<pal::string_t> &files)
 {
-    if (m_enabled)
+    auto instance = std::make_shared<breadcrumb_writer_t>(files);
```

Comment:
While you're cleaning this up, would it be possible to get rid of `begin_write()` and `end_write()`, and use only the constructor/destructor of `breadcrumb_writer_t`?  If that's possible, then `make_shared()` can be called in `run_app_for_context()`, and the branch to determine if `end_write()` has to be called can be junked. This would simplify this class a little bit.

Another thing that might be possible here is to get rid of `breadcrumb_writer_t` entirely and capture the breadcrumbs with a lambda passed directly to `std::thread`.  I like the call to `swap()` to own the breadcrumbs map, though, and I don't know if there's a good way to do something similar without it.

Reply:
Something like this could potentially simplify things: https://gist.github.com/lpereira/48403e8540524b7854743d68b6dd879d